### DISCUSSION TITLE
calling __ISB(); in vPortSuppressTicksAndSleep will cause hardfault in STM32L031

### DISCUSSION THREAD

**Discussion by Scarittagle** [jefftenney/LPTIM-Tick#18] [Q&A]
looks like a hardfault will occur whenever I called `osDelay()` or anything that uses it (for example: `osThreadFlagsWait()`) .

when tracing it down through debugger, it was stuck in [here](https://github.com/jefftenney/LPTIM-Tick/blob/6dc0a7d6ebcb2d449efff60f0ba85bca24fdb87c/Core/Src/lptimTick.c#L482C13-L482C103)

ulp.c does not seems to be related, this happens with and without ulp.c enabled.

Because the project was generated by CubeMX, The FreeRTOS kernel has been replaced with version 10.4.0.

This problem does not occur in my STM32G0 implementation.

**Comment by jefftenney** [comment]
Hi @Scarittagle,

I don't have any 'L0 hardware to test on.  Does this hardfault occur even when using basic sleep only -- no STOP mode?

**Reply by Scarittagle** [reply]
@jefftenney ,

Yes, it will occur when using basic sleep. I have to disable the Debug in Low Power Mode Option in the IDE to get it to work.

**Reply by Olstyle** [reply]
I have seen this with -o0 optimization. For some reason GCC decided to add NOPs before and after the inline function, which contradicts the original purpose of those barriers.
Try to have a look at the assembly in those lines.
